autode/pes/reactive.py

```diff
@@ -0,0 +1,351 @@
+        discreet potential energy surface between the initial species
+        the 'reactant' state, to a new product state. Points are identified
+        based on graph isomorphisms
+
+        -----------------------------------------------------------------------
+        Arguments:
+            product: Product (hopefully) on the surface
+
+        Yields:
+            (autode.transition_states.ts_guess.TSguess): TS guess
+
+        Returns:
+            (StopIteration): If there are no suitable TS guesses
+        """
+        reactant = self._species
+
+        if reactant.graph is None or product.graph is None:
+            logger.warning('Products or reactants did not have a defined '
+                           'graph, thus the MEP could not be traversed')
+            return StopIteration
+
+        product_point = self._point_with_isomorphic_graph_to(product)
+
+        if product_point is None:
+            logger.warning('Could not find any point on the surface that had '
+                           f'the same connectivity as {product}')
+            return StopIteration
+
+        ts_point = peak_point(energies=self._energies,
+                              point1=self.origin,
+                              point2=product_point)
+
+        species = self._species.new_species()
+        species.coordinates = self._coordinates[ts_point]
+
+        yield TSguess.from_species(species)
+
+    def _point_with_isomorphic_graph_to(self,
+                                        species: 'autode.species.Species'
+                                        ) -> Optional[Tuple]:
+        """
+        Find a point on this surface that is graph-isomorphic to a particular
+        species. Attempt to return the lowest energy point.
+
+        -----------------------------------------------------------------------
+        Arguments:
+            species:
+
+        Returns:
+            (tuple(int, ..) | None):
+        """
+        isomorphic_points = []
+
+        for point in self._points():
+
+            mol = self._species.new_species()
+            mol.coordinates = self._coordinates[point]
+            mol.reset_graph()
+
+            if mol.has_same_connectivity_as(species):
+                isomorphic_points.append(point)
+
+        if len(isomorphic_points) == 0:
+            logger.warning('No isomorphic points found')
+            return None
+
+        min_idx = np.argmin([self._energies[p] for p in isomorphic_points])
+        return isomorphic_points[min_idx]
+
+    def _set_hessian(self, point: Tuple) -> None:
+        """
+        Set the Hessian for a particular point in the surface, evaluated
+        using finite differences
+
+        -----------------------------------------------------------------------
+        Arguments:
+            point:
+        """
+        hessian = self._hessians[point]
+
+        for i in range(self.ndim):
+            for j in range(i, self.ndim):
```

Comment:
do we want a `i > j` check here?

Reply:
i in [0, n) and j in [i, n) means it's not needed(?)

Reply:
yep good point, didn't notive that